# Numeric Types 

----
## Numbers

Integer Numbers. Python Type: int
<br>
Rational Numbers. Python Type: fractions.Fraction
<br>
Real Numbers. Python Type: float decimal.Decimal
<br>
Complex Numbers. Python Type: complex
<br>
Boolean. 

----
## Integer: Data Types

Integers are represented internally using base-2 (binary) digits, not decimal.

The int object uses a variable number of bits.
<br>
Can use 4 bytes (32 bits), 8 bytes (64 bits), 12 bytes (96 bits), etc.
<br>
Since ints are actually objects, there is a further fixed overhead per integer
<br>
Theoretically limited only by the amount of memeory available.

In [2]:
print(type(100))

<class 'int'>


In [3]:
import sys

In [4]:
sys.getsizeof(0)

24

Allow us to find out how much memory is used to store a particular value.
<br>
The result is 24 bytes.

In [5]:
sys.getsizeof(1)

28

In [6]:
sys.getsizeof(2**1000)

160

In [7]:
(160 - 24) * 8

1088

We used 1088 bits to store 160 bytes

In [8]:
import time

In [9]:
def calc(a):
    for i in range(10000000):
        a * 2

In [10]:
start = time.perf_counter()
calc(10)
end = time.perf_counter()
print(end - start)

0.5089969329999349


In [12]:
start = time.perf_counter()
calc(2**100)
end = time.perf_counter()
print(end - start)

0.8276703349999934


In [13]:
start = time.perf_counter()
calc(2**10000)
end = time.perf_counter()
print(end - start)

5.805058163000012


## Integers: Operations

First, let's revisit long integer division
<br>
155 = numerator
<br>
////
<br>
4 = denominator
<br>
155 / 4 = 38 with remainder 3

Put another way:
<br>
155 = 4 * 38 + 3

In [16]:
155 // 4

38

Is called the floor division

In [17]:
155 % 4

3

Is called the modulo operator (mod)

155 = 4 * (155 // 4) + (155 % 4)

These operators always satisfy the operation above, which is the next equation:

n = d * (n // d) + (n % d)

Let's see the floor division with negative numbers

The floor of a real number <b>a</b> is the largest (in the standard number order) integer <= <b>a</b>

------O------o----O
<br>
---- -4 --- -3.1 -3
<br>
Tip: Floor, in a line where we go from 0 to 100 and 0 to -100, floor will always go for the integer number from the left.

In [23]:
-3.1//1

-4.0

So, floor iis not quite the same as truncation

In [24]:
a = 135
b = 4

In [25]:
135 / 4

33.75

In [26]:
135 // 4 

33

In [29]:
-135 // 4

-34

In [27]:
135 % 4

3

In [30]:
-135 % 4

1

With the equation:

In [28]:
4 * (135 // 4) + (135 % 4)

135

In [31]:
4 * (-135 // 4) + (-135 % 4)

-135

In [37]:
type(1 + 1)

int

In [38]:
type(2 * 4)

int

In [39]:
type(4 - 10)

int

In [40]:
type(3 ** 6)

int

In [41]:
type(2 / 3)

float

In [42]:
type( 2 // 3)

int

In [43]:
type(10/2)

float

In [53]:
import math

In [54]:
math.floor(3.15)

3

In [55]:
math.floor(3.99999)

3

In [65]:
math.floor(-144.41)

-145

In [68]:
math.floor(-3.0000000000000001)

-3

Float has a limited position in Python

In [69]:
a = 33
b = 16

In [72]:
print(a/b)
print(a // b)
print(math.floor(a/b))

2.0625
2
2


In [73]:
a = -33
b = 16

In [74]:
print(a/b)
print(a // b)
print(math.floor(a/b))

-2.0625
-3
-3


In [75]:
a = -33
b = 16
print(a/b)
print(a // b)
print(math.floor(a/b))
print(math.trunc(a/b))

-2.0625
-3
-3
-2


In [76]:
a = b * (a//b) + (a%b)

In [78]:
a = 13
b = 4
print(f'{a}/{b} = {a/b}')
print(f'{a}//{b} = {a//b}')
print(f'{a}%{b} = {a%b}')
print(a == b * (a//b) + (a%b))

13/4 = 3.25
13//4 = 3
13%4 = 1
True


In [79]:
a = -13
b = 4
print(f'{a}/{b} = {a/b}')
print(f'{a}//{b} = {a//b}')
print(f'{a}%{b} = {a%b}')
print(a == b * (a//b) + (a%b))

-13/4 = -3.25
-13//4 = -4
-13%4 = 3
True


In [80]:
a = 13
b = -4
print(f'{a}/{b} = {a/b}')
print(f'{a}//{b} = {a//b}')
print(f'{a}%{b} = {a%b}')
print(a == b * (a//b) + (a%b))

13/-4 = -3.25
13//-4 = -4
13%-4 = -3
True


In [81]:
a = -13
b = -4
print(f'{a}/{b} = {a/b}')
print(f'{a}//{b} = {a//b}')
print(f'{a}%{b} = {a%b}')
print(a == b * (a//b) + (a%b))

-13/-4 = 3.25
-13//-4 = 3
-13%-4 = -1
True


---
## Integers: Constructors and Bases

The <b>int</b> class provides multiple constructors
<br>
a = int(10)
<br>
a = int(-10)
<br>

Other (numerical) data types are also supported in the argument of the <b>int</b> constructor:
<br>
a = int(10.9) -> truncation a = 10
<br>
a = int(-10.9) -> truncation a = -10
<br>
a = int(True) -> a = 1
<br>
a = int(Decimal("10.9")) -> truncation a = 10

As well as strings (that can be parsed to a number)
<br>
a = int("10") -> a = 10

### - Number Base

int("123") -> (123)base(10)
<br>
When used with a string, constructor has an <b>optional second</b> parameter: <b>base</b> (2 <= base <= 36)
<br>
If base is not specified, the default is base 10 - as the example above.
<br>

int("1010", 2) -> (10)base(10) 

<br>

int("1010", base=2) -> (10)base(10)

<br>

int("A12F", base=16) -> (41263)base(10)

<br>

int("a12f", base=16) -> (41263)base(10)

<br>

int("1010", base=2) -> (10)base(10)

<br>

int("534", base=8) -> (348)base(10)

<br>

int("A", base=11) -> (10)base(10)

<br>

int("B", 11) ValueError: invalid literal..

### - Reverse Process: changing an integer from base 10 to another base
<br>

built-in functions:

<br>

- bin()  bin(10) -> '0b1010'
- oct()  oct(10) -> '0o12'
- hex()  hex(10) -> '0xa'

<br>

The prefixes in the strings help document the base of the number: int('0xA', 16) -> (10)base(10)
<br>
These prefixes are consistent with literal integers using a base prefix (no strings attached)

<br>
- a = 0b1010 a -> 10
- a = 0o12 a -> 10
- a = 0xA a -> 10
<br>

### - What about other bases? Custom code

<b>n</b>: number(base 10)
<br>
<b>b</b>: base (target base)
    

n = b * (n // b) + n % b
<br>
-> n= (n // b) * b + n % b

In [2]:
n = 232
b = 5

We are giving the number a different representation of a number. It's always a binary number insider our computer

We want to know what number in base 5 is 232.
<br>
Could be:
- 5^3
- 5^2
- 5^1
- 5^0

        232 = (232 // 5) * 5 + 232 % 5 = 46 * 5 + 2
        <br>
        = [46 * 5^1] + [2 * 5^0]      -> 5^1 = 46 which is too big for base 5
        <br>
        = [((46 // 5) * 5 + 46 % 5) * 5^1] + [2 * 5^0]
        <br>
        = [(9 * 5 + 1) * 5^1] + [2 * 5^0]     -> 5^2 = 9 which is too big for base 5
        <br>
        = [((9 // 5) * 5 + 9 % 5) * 5^2] + [1 + 5^1] + [2 * 5^0]
        <br>
        = [(1 * 5 + 4) * 5 ^2 ] + [1 * 5^1] + [2 * 5^0]
        <br>
        = [1 * 5^3] + [4 * 5^2] + [1 * 5^1] + [2 * 5^0]    -> 5^3 = 1 whis is ok.
        
        <br>
        The above is: 
        div + 3rd mod + 2nd mod + 1st mod
        
        - 5^3 = 1
        - 5^2 = 4
        - 5^1 = 1
        - 5^0 = 2

### - Base Change Algorithm
n = base-10 number (>=0)   b = base (>=2)
<br>
if b<2 or n<0: raise exception
<br>
if n == 0: return [0]
<br>
digits = [ ]
<br>
while n > 0:
<br>
    m = n % b
    <br>
    n = n//b
    <br>
    digits.insert(0,m)
    
<br>
Example: n = 232, b = 5
<br>
digits = [1,4,1,2]
<br>
Example 2: n = 1485, b = 16
<br>
digits = [5,12,13]
<br>

This algorithm retursn a list of the digits in the specified base b (a representation of nbase10 in base b)
<br>
Usually we want to return an encoded number where digits higher than 9 use letters such as A-Z
<br>
We simply need to decide what character to use for the various digits in the base


### - Encodings
Typically, we use 0-9 and A-Z for digits required in bases higher than 10.

But we don't have to use letters or even statndar 0-9 digits to encode our number.

We just need to map between the digits in our number, to a character of our choice.

0 ->0

1 -> 1

0 -> a
...
36 -> *

Python uses 0-9 and a-z(case insenstive) and is therefore limited to base <= 36

The simples way to do this given a list of digits to encode, is to create a string with as many characaters as needed, and use their index(ordinal pos) for our encoding map.

base b (>=2)
<br>
map = '...' (of length b)
<br>
digits = [...]
<br>
encoding = map[digits[0]] + map[digits[1]] + ...
<br>
Example: Base 12

map = '0123456789ABC'
<br>
digits = [4,11,3,12]
<br>
encoding = '4B3C'


### - Encoding Algorithm
digits = [...]
<br>
map = '...'
<br>
encoding = ''
<br>
for d in digits:
<br>
    encoding += map[d]
    
or, more simply

enconding = ''.join([map[d] for d in digits])

In [3]:
type(10)

int

In [4]:
help(int)

Help on class int in module builtins:

class int(object)
 |  int([x]) -> integer
 |  int(x, base=10) -> integer
 |  
 |  Convert a number or string to an integer, or return 0 if no arguments
 |  are given.  If x is a number, return x.__int__().  For floating point
 |  numbers, this truncates towards zero.
 |  
 |  If x is not a number or if base is given, then x must be a string,
 |  bytes, or bytearray instance representing an integer literal in the
 |  given base.  The literal can be preceded by '+' or '-' and be surrounded
 |  by whitespace.  The base defaults to 10.  Valid bases are 0 and 2-36.
 |  Base 0 means to interpret the base from the string as an integer literal.
 |  >>> int('0b100', base=0)
 |  4
 |  
 |  Built-in subclasses:
 |      bool
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __bool__(self, /)
 |      self != 

In [6]:
int(10.5)

10

In [7]:
int('23')

23

In [8]:
int(True)

1

In [9]:
int(False)

0

In [10]:
import fractions

In [12]:
a = fractions.Fraction(22,7)

In [13]:
a

Fraction(22, 7)

In [14]:
print(a)

22/7


In [15]:
float(a)

3.142857142857143

In [16]:
int(a)

3

In [17]:
int('12345')

12345

In [18]:
int('101', 2)

5

In [19]:
int('FF', 16)

255

In [20]:
int('ff', 16)

255

In [21]:
int('A', 11)

10

In [22]:
int('B', 11)

ValueError: invalid literal for int() with base 11: 'B'

In [23]:
bin(10)

'0b1010'

In [24]:
bin(5)

'0b101'

In [25]:
oct(10)

'0o12'

In [26]:
hex(255)

'0xff'

In [27]:
a = int('101',2)
#instead, we can say
b = 0b101

In [28]:
a

5

In [29]:
b

5

In [50]:
def from_base10(n,b): #convert from base 10 to any other base
    if b < 2:
        raise ValueError('Base b must be >= 2')
    if n < 0:
        raise ValueError('Number n must be >= 0')
    if n == 0:
        return [0]
    digits = []
    while n > 0:
        #an option:
#        m,n = n % b, n // b
        #another equivalent way
        n,m = divmod(n,b)
#        print(f'n: {n}, m: {m}')
        digits.insert(0,m)
 #       print(digits)
    return digits

In [51]:
from_base10(10,2)

[1, 0, 1, 0]

In [52]:
from_base10(255,16)

[15, 15]

In [53]:
def encode(digits, digit_map):
    if max(digits) >= len(digit_map):
        raise ValueError('digit_map is not long enought to encode the digits')
    return ''.join([digit_map[d] for d in digits])

In [54]:
encode([15,15], '0123456789ABCDEF')

'FF'

In [58]:
def rebase_from10(number, base):
    digit_map = '0123456789ABCDEFGHIJKLMNOPRSTUVWXYZ'
    if base < 2 or base > 36:
        raise ValueError('Invalid base: 2 <= base <= 36')
    sign = -1 if number < 0 else 1
    number *= sign
    digits = from_base10(number,base)
    encoding = encode(digits, digit_map)
    return '-'+encoding if sign == -1 else encoding

In [59]:
e = rebase_from10(10,2)
print(e)
print(int(e,2))

1010
10


In [60]:
e = rebase_from10(3451, 16)
print(e)
print(int(e,16))

D7B
3451


In [63]:
e = rebase_from10(-314, 2)
print(e)
print(int(e,2))

-100111010
-314


----
## Rational numbers

Rational numbers are fractions of integer numbers

Any real number with a finite number of digits after the decimal point is also a rational nummber

So, 8.3 / 4 is also rational

as is 8.3/1.4



Rational numbers can be repreented in Python using Fraction class in the fractions module

Fractions are automatically reduced

Negative sign, if any, is always attached to the numerator
<br>
Fraction(1,-4) - > Fraction(-1,4)


Fraction(numerator = 0, denominator = 1) *default values 0 and 1


Fraction('10') = Fraction (10,1)
<br>
Fraaction('0.25') = Fraction(1,1)
<br>
Fraction('1/6') = Fraction(1,6)

Standard artihmetic operators are supported: +,-,*,/ 
and result in Fraction objects as well

We can access the numerator and denomitaro of Fraction objetcts:

x = Fraction(22,7)
<br>
x.numerator = 22
<br>
x.denominator = 7

Every number we can store in the computer is finite, also <b>float</b> objects have <b>finite</b> precision.

Any float object can be written as a fraction.

Fraction(0.75) - > Fraction(3,4)

### - Irrational numbers

x = Fraction(math.pi) - > Fraction(884279..., 2814749...)
<br>
Event though <b>π</b> and <b>√2</b>  irrational:
<br>
- internally represented as floats
- finite precision real numbers
- expressible as a rational numbers

**but is an aproxximation**

### - Converting a float to a Fraction has an important caveat

What happens when we do not have the exact float representation?
<br>
Example: 3/10 = 0.33333333....

So, we are going to have an approximation.

### Constraining the denominator
Given a Fraction object, we can find an approximate equivalent fraction with a constrained denominator using *limit_denominator(max_denominator=100000)* instance method.

i.e. finds the closes rational (which coud be a precisely equal) with a denominator that does not exceed *max_denominator*

In [64]:
from fractions import Fraction

In [66]:
help(fractions)

Help on module fractions:

NAME
    fractions - Fraction, infinite-precision, real numbers.

CLASSES
    numbers.Rational(numbers.Real)
        Fraction
    
    class Fraction(numbers.Rational)
     |  Fraction(numerator=0, denominator=None, *, _normalize=True)
     |  
     |  This class implements rational numbers.
     |  
     |  In the two-argument form of the constructor, Fraction(8, 6) will
     |  produce a rational number equivalent to 4/3. Both arguments must
     |  be Rational. The numerator defaults to 0 and the denominator
     |  defaults to 1 so that Fraction(3) == 3 and Fraction() == 0.
     |  
     |  Fractions can also be constructed from:
     |  
     |    - numeric strings similar to those accepted by the
     |      float constructor (for example, '-2.3' or '1e10')
     |  
     |    - strings of the form '123/456'
     |  
     |    - float and Decimal instances
     |  
     |    - other Rational instances (including integers)
     |  
     |  Method resoluti

In [67]:
Fraction(1)

Fraction(1, 1)

In [68]:
Fraction(denominator = 1, numerator = 2)

Fraction(2, 1)

In [69]:
Fraction('1/6')

Fraction(1, 6)

In [70]:
Fraction(numerator = 1, denominator = 2)

Fraction(1, 2)

In [71]:
Fraction(0.125)

Fraction(1, 8)

In [72]:
Fraction(2,3) + Fraction(3,4)

Fraction(17, 12)

In [77]:
x = Fraction(0.125)
print(x.numerator,'/',x.denominator)

1 / 8


In [78]:
import math
x = Fraction(math.pi)
x

Fraction(884279719003555, 281474976710656)

In [79]:
float(x)

3.141592653589793

π is not rational, therefore an approximation of real numbers is represented

In [80]:
y = Fraction(math.sqrt(2))

In [81]:
y

Fraction(6369051672525773, 4503599627370496)

In [82]:
float(y)

1.4142135623730951

In [83]:
a = 0.125
b = 0.3

In [84]:
print(a)

0.125


In [85]:
print(b)

0.3


In [86]:
Fraction(a)

Fraction(1, 8)

In [87]:
Fraction(b)

Fraction(5404319552844595, 18014398509481984)

Python is trying to give us a nice dispplay, which is not in reality is not stored

In [90]:
format(b, '0.5f')

'0.30000'

In [91]:
format(b, '0.15f')

'0.300000000000000'

In [92]:
format(b, '0.25f')

'0.2999999999999999888977698'

Now we can see that 0.3 is not actually stored as 0.3

That's why that we get an approximation with Fraction

In [94]:
Fraction(b)

Fraction(5404319552844595, 18014398509481984)

In [95]:
x = Fraction(0.3)
x.limit_denominator(10) #we dont want a denominator greater than 10

Fraction(3, 10)

In [96]:
x = Fraction(math.pi)
x

Fraction(884279719003555, 281474976710656)

In [97]:
float(x)

3.141592653589793

In [98]:
x.limit_denominator(10)

Fraction(22, 7)

The closest fraction we get to x nad limited to 10

In [99]:
22/7

3.142857142857143

In [100]:
x.limit_denominator(100)

Fraction(311, 99)

In [101]:
311/99

3.1414141414141414

---
## Floats: Internal representation

The **float** class is Python's default implementation for representing real numbers.

The Python (Cpython) float is implemented using the C double type which (usually!) implementes the IEEE 754 double-precision binary float, also called binary64.

The float uses a fixed number of bytes -> 8 bytes = 64 bits.

These 64 bits are used up as follows:

sign - > 1 bit

exponent -> 11 bits - > range[-1022, 1023]

1.5e-5 -> 1.5 x 10^-5

significant digits - > for simplicity,all digits except leading and trailing zeros.

1.2345 has 5 significant digits.

1234.5 has also 5 significan digits.

12345000000 has also 5 signficant digits, it is just 12345 times 1000000.





### - Representation: decimal
Numbers can be represented as base-10 integers and fractions:

0.75 - >  7/10 + 5/100 - > 7 x 10^-1 + 5 x 10^-2
<br>
This means there are 2 significant digits.

0.256 - > 2/10 + 5/100 + 6/1000 -> 2 x 10^-1 + 5 x 10^-2 + 6 x 10 ^-3
<br>
This means there are 3 significant digits.

123.456 - > 1 x 100 + 2 x 10 + 3 x 1 + 4/10 + 5/100 + 6/1000 <br>
          - > 1 x 10^2 + 2 x 10^1 + 3 x 10^0 + 4 x 10 ^-1 + 5 x 10^-2 + 6 x 10^-3
<br>
This means there are 6 significant digits


In general: d = ∑n, i = -m, di x 10^i

Some number cannot be represented using a finite number of terms.

Obviously numbers such as :
- π = 3.14159...
- √2 = 1.4142...

but even some rational numbers : 
- 1/3 = 0.333333...

### - Representation: binary
Numbers in a computer are represented using bits, not decimal digits

instead of powers of 10, we need to use powers of 2

(0.11)base(2) = (1/2 + 1/4)base(10) = (0.5 + 0.25)base(10) = (0.75)base(10)
<br> ========== (1 x 2^-1 + 1 x 2^-2)

This representation is very similar to the one we use with decimal numbers but instead of using powers of 10, we use powers of 2 a binary representation.

In general: d = ∑n, i = -m, di x 10^i

The same problem that occurs when trying to represent 1/3 using a decimal exapansion also happens when trying to represent certain numbers using binary expansion.

0.1 = 1/10

(0.1)base(10) = (0.0 0011 0011 0011 ...)base(2)
<br>
= 0/2 + 0/4 + 0/8 + 1/16 + 1/32...

So, some number that do have a finite decimal representation do not have a finite binary representation, and some do.

(0.75)base(10) = (0.11)base(2)      finite
<br>
(0.8125)base(10) = (0.1101)base(2)  finite  exact float representation
<br>
(0.1)base(10) = (0 0011 0011 0011...)base(2)   infinite approxx float representation



In [102]:
help(float)

Help on class float in module builtins:

class float(object)
 |  float(x=0, /)
 |  
 |  Convert a string or number to a floating point number, if possible.
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __bool__(self, /)
 |      self != 0
 |  
 |  __ceil__(self, /)
 |      Return the ceiling as an Integral.
 |  
 |  __divmod__(self, value, /)
 |      Return divmod(self, value).
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __float__(self, /)
 |      float(self)
 |  
 |  __floor__(self, /)
 |      Return the floor as an Integral.
 |  
 |  __floordiv__(self, value, /)
 |      Return self//value.
 |  
 |  __format__(self, format_spec, /)
 |      Formats the float according to format_spec.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getnewargs__(self, /)
 |  
 |  __gt__

In [103]:
float(10)

10.0

In [104]:
float(10.0)

10.0

In [105]:
float('10.4')

10.4

In [106]:
float('22/7')

ValueError: could not convert string to float: '22/7'

In [108]:
from fractions import Fraction

In [109]:
a = Fraction('22/7')
float(a)

3.142857142857143

In [110]:
print(0.1)

0.1


The problem is, as the previous example with 0.3, Python is showing a different display. Behind the scenes, this is not the real value.

In [111]:
format(0.1, '.25f')

'0.1000000000000000055511151'

In [112]:
print(0.125)

0.125


In [113]:
print(1/8)

0.125


In [114]:
format(0.125, '0.25f')

'0.1250000000000000000000000'

In [115]:
a = 0.1 + 0.1 + 0.1

In [116]:
b = 0.3

In [117]:
a == b

False

In [118]:
format(a, '0.25f')

'0.3000000000000000444089210'

In [119]:
format(b, '0.25f')

'0.2999999999999999888977698'

**0.1** and **0.3** is exact in base 10, but not in binary representation, not in base 2

----
## Floats: Equality Testing

In [6]:
x = 0.1
format(x, '0.25f')

'0.1000000000000000055511151'

In [9]:
x = 0.125 + 0.125 + 0.125
y = 0.375

In [10]:
x == y

True

In [11]:
x = 0.1 + 0.1 + 0.1
y = 0.3

In [12]:
x == y

False

One option is to round

In [13]:
print(format(x,'.25f'))
print(format(y,'.25f'))

0.3000000000000000444089210
0.2999999999999999888977698


In [14]:
round(x,3) == round(y,3)

True

The problem is that round is testing how close the numbers are within an absolute tolerance.

That has problems, that absolutes tolerance can be a problem when comparing larger numbers.

In [17]:
x = 10000.01
y = 10000.02
y / x
#Delta = 0.01

1.000000999999

In [18]:
x = 0.01
y = 0.02
y / x
#Delta = 0.01

2.0

This is the issue, which we consider closest. Absolute tolerance could be a problem

In [19]:
round(x,1) == round(y,1)

True

In [20]:
x = 10000.01
y = 10000.02
round(x,1) == round(y,1)

True

For that reason, let's use the *isclose* method

In [21]:
from math import isclose

In [22]:
help(isclose)

Help on built-in function isclose in module math:

isclose(a, b, *, rel_tol=1e-09, abs_tol=0.0)
    Determine whether two floating point numbers are close in value.
    
      rel_tol
        maximum difference for being considered "close", relative to the
        magnitude of the input values
      abs_tol
        maximum difference for being considered "close", regardless of the
        magnitude of the input values
    
    Return True if a is close in value to b, and False otherwise.
    
    For the values to be considered close, the difference between them
    must be smaller than at least one of the tolerances.
    
    -inf, inf and NaN behave similarly to the IEEE 754 Standard.  That
    is, NaN is not close to anything, even itself.  inf and -inf are
    only close to themselves.



In [23]:
x = 0.1 + 0.1 + 0.1
y = 0.3

In [24]:
isclose(x,y)

True

In [25]:
x == y

False

In [28]:
x = 123456789.01
y = 123456789.02
isclose(x,y,rel_tol=0.01)

True

In [29]:
x = 0.01
y = 0.02
isclose(x,y,rel_tol=0.01)

False

In [30]:
x = 0.0000001
y = 0.0000002
isclose(x,y, rel_tol=0.01)

False

In this case, we want to use the absolute tolerance.

In [32]:
isclose(x,y, rel_tol = 0.01, abs_tol=0.01)

True

In [33]:
x = 0.0000001
y = 0.0000002

a = 123456789.01
b = 123456789.02

print(isclose(x,y, abs_tol= 0.0001, rel_tol=0.01))
print(isclose(a,b, abs_tol=0.0001, rel_tol=0.01))

True
True


For small numbers we want absolute tolerances.

For large numbers we want relative tolerances.

-----
## Floats: Coercing to integers

### - Truncation
Truncating a float simply returns the integer portion of the number.
i.e. ignores everyhting after the decimal point.


###  + int constructor
The int constructor accepts a float.

Uses truncation ewhe casting the float to a int

### - Floor
The floor of a number is the largest integer less than (or equal to) the number.
Recall also our discussion on integer division - aka floor division: //

We defined floor division in combination with the mod operatio **n = d * (n // d) + (n % d))**

But int fact, floor division defined that way yields the same result as taking the floor of the floating point division

**a // b = floor(a / b)**

For positive numbers, Floor and Truncation are equivalent

### - Ceiling
The ceiling of a number is the smalles integer **greater** than (or equal to ) the number

----
## Floats: Rounding

n = 0

round to the closest multiple of 10^-0 = 1

-1--x----2

x = 1.23

Round will round to the closest, which is 1


++++++++++++++++++++++++++++++++++++++++++++++++++

n = 1

round to the closest multiple of 10^-1 = 0.1

x = 1.23

-1.2--x----1.3-

round(1.23,1) = 1.2

n = -1

round to the closest multiple 10^-(-1) = 10

x = 18.2

-10-----x--20-

round(18.2, -1) = 20

### - Ties

x = 1.25

-1.2-----x--1.3-

We probably would expect **round(1.25,1)** to be 1.3 because there is no closest value

Similarly, we would expect **round(-1.25,1)** to result in -1.3

***We can say we are rounding away from zero***

This type of roundis is called **rounding to neares, with ties away from zero**

But in fact:  round(1.25,1) - > 1.2 towards 0

___________  round(1.35, 1) - > 1.4 away from 0

___________  round(-1.25, 1) - > -1.2 towards 0

___________  round(-1.35, 1) - > -1.4 away from 0


### - Bankers Rounding

It rounds to the nearest value, with ties rounded to the nearest value with an **even** least significant digit.

In case we have a tie:

x = 1.25

1.2----x----1.3


from 1.2, 2 is even, so round(1.25,1) - > 1.2

x = 1.35

1.3----x----1.4

Since 4 is even, so round(1.35,1) - > 1.4


****we pick the last significant digit****



### Why Banker's Rounding?

Less biased rounding than ties away from zero

Consider averaging three numbers, and averaging the rounded values of each.

If you really insist on rounding away from zero...

One common (and partially incorrect) way to round to nearest unit that often comes up on the web is:

int(x + 0.5)

but, it does not work for negative numbers.

-10.3 -> int(-10.3 + 0.5) = int(-9.8) = -9

If you really insist on rounding away from zero...

The correct way to do it:

sign(x) * int(abs(x)+ 0.5))

_sign is a function where it return -1 or 1_

In [33]:
help(round)

Help on built-in function round in module builtins:

round(number, ndigits=None)
    Round a number to a given precision in decimal digits.
    
    The return value is an integer if ndigits is omitted or None.  Otherwise
    the return value has the same type as the number.  ndigits may be negative.



In [37]:
a = round(1.9)
a, type(a)

(2, int)

In [38]:
a = round(1.9, 0)
a, type(a)

(2.0, float)

n > 0

In [45]:
round(1.8888,3), round(1.8888,2), round(1.8888,1), round(1.8888,0),

(1.889, 1.89, 1.9, 2.0)

n < 0

In [52]:
round(888.88,1), round(888.88,0), round(888.88,-1), round(888.88,-2), round(888.88,-3), round(888.88,-4)

(888.9, 889.0, 890.0, 900.0, 1000.0, 0.0)

In [65]:
round(9800,-4)

10000

#### Ties

In [68]:
round(1.25,1)

1.2

In [70]:
round(1.35,1)

1.4

In [72]:
round(-1.25, 1), round(-1.35, 1)

(-1.2, -1.4)

In [83]:
def _round(x):
    #Onlly rounds integer values
    from math import copysign
    return int(x + 0.5 * copysign(1,x))

In [84]:
round(1.5), _round(1.5)

(2, 2)

In [85]:
round(-1.5), _round(-1.5)

(-2, -2)

In [86]:
round(2.5), _round(2.5)

(2, 3)

In [90]:
round(-2.5), _round(-2.5)

(-2, -3)

Help on built-in function round in module builtins:

round(number, ndigits=None)
    Round a number to a given precision in decimal digits.
    
    The return value is an integer if ndigits is omitted or None.  Otherwise
    the return value has the same type as the number.  ndigits may be negative.



---
## Decimals

In [1]:
import decimal
from decimal import Decimal

In [2]:
decimal.getcontext()

Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

In [5]:
decimal.getcontext().rounding

'ROUND_HALF_EVEN'

In [8]:
decimal.getcontext().prec = 6

In [9]:
decimal.getcontext()

Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

In [19]:
g_ctx = decimal.getcontext()
type(g_ctx)

decimal.Context

In [20]:
g_ctx.rounding = decimal.ROUND_HALF_UP

In [21]:
type(decimal.ROUND_HALF_UP)

str

In [22]:
decimal.getcontext()

Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

In [24]:
g_ctx.prec = 28
g_ctx.rounding = decimal.ROUND_HALF_EVEN

In [25]:
decimal.getcontext()

Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

In [26]:
type(decimal.localcontext())

decimal.ContextManager

In [27]:
type(decimal.getcontext())

decimal.Context

In [31]:
x = Decimal('1.25')
y = Decimal('1.35')

In [34]:
with decimal.localcontext() as ctx:
    ctx.prec = 6
    ctx.rounding = decimal.ROUND_HALF_UP
    print(round(x,1))
    print(round(y,1))

#The context in the module level is to ROUND_HALF_EVEN
print(round(x,1))
print(round(y,1))

1.3
1.4
1.2
1.4


-----
## Decimals: Constructors and Contexts

In [42]:
import decimal
from decimal import Decimal

In [43]:
help(Decimal)

Help on class Decimal in module decimal:

class Decimal(builtins.object)
 |  Decimal(value='0', context=None)
 |  
 |  Construct a new Decimal object. 'value' can be an integer, string, tuple,
 |  or another Decimal object. If no value is given, return Decimal('0'). The
 |  context does not affect the conversion and is only passed to determine if
 |  the InvalidOperation trap is active.
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __bool__(self, /)
 |      self != 0
 |  
 |  __ceil__(...)
 |  
 |  __complex__(...)
 |  
 |  __copy__(...)
 |  
 |  __deepcopy__(...)
 |  
 |  __divmod__(self, value, /)
 |      Return divmod(self, value).
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __float__(self, /)
 |      float(self)
 |  
 |  __floor__(...)
 |  
 |  __floordiv__(self, value, /)
 |      Return self//value.
 |  
 |  __format__(...)
 |      Default object formatter.
 |

In [44]:
Decimal(10)

Decimal('10')

In [45]:
Decimal('10.1')

Decimal('10.1')

In [46]:

Decimal('-3.1415')

Decimal('-3.1415')

In [47]:
#With tuples
t = (0, (3,1,4,1,5), -4)

In [48]:
Decimal(t)

Decimal('3.1415')

In [50]:
#0 for positive
Decimal((0, (3,1,4,1,5), -4))

Decimal('3.1415')

In [51]:
#1 for negative
Decimal((1, (3,1,4,1,5), -4))

Decimal('-3.1415')

In [54]:

Decimal((1, (3,1,4,1,5), -2))

Decimal('-314.15')

In [55]:
format(0.1, '.25f')

'0.1000000000000000055511151'

In [56]:
Decimal(0.1)

Decimal('0.1000000000000000055511151231257827021181583404541015625')

In case you do not want the exact number, just get in the constructor a string

In [57]:
Decimal('0.1')

Decimal('0.1')

In [58]:
Decimal(0.1) == Decimal('0.1')

False

In [59]:
decimal.getcontext()

Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

In [62]:
decimal.getcontext().prec = 2

In [63]:
a = Decimal('0.12345')
b = Decimal('0.12345')

In [64]:
a,b

(Decimal('0.12345'), Decimal('0.12345'))

In [65]:
0.12345 + 0.12345

0.2469

The precision did affect the sum, as we can see below:

In [70]:
with decimal.localcontext() as ctx:
    ctx.prec = 2
    result = a + b
    print('result with local context: ', result)
print('result without local context: ', a+b)

result with local context:  0.25
result without local context:  0.24690


-----
## Decimals: Math Operations

Some artihmetic operator don't work the same as floats or integers.

**//** and **%** and also **divmod()**

For integers, the **//** operator performs floor division - > a // b = floor(a/b)

For Decimales, however, it performs truncated division - >  a // b = trunc(a/b)

In [74]:
import decimal
from decimal import Decimal

// and %

n = d * (n//d) + (n%d)

With integers

In [77]:
x = 10
y = 3
print(x//y, x%y)
print(divmod(x,y))
print(x == y * (x//y) + (x%y))

3 1
(3, 1)
True


With decimals

In [78]:
x = Decimal(10)
y = Decimal(3)
print(x//y, x%y)
print(divmod(x,y))
print(x == y * (x//y) + (x%y))

3 1
(Decimal('3'), Decimal('1'))
True


With negative numbers

In [79]:
x = -10
y = 3
print(x//y, x%y)
print(divmod(x,y))
print(x == y * (x//y) + (x%y))

-4 2
(-4, 2)
True


In [80]:
x = Decimal(-10)
y = Decimal(3)
print(x//y, x%y)
print(divmod(x,y))
print(x == y * (x//y) + (x%y))

-3 -1
(Decimal('-3'), Decimal('-1'))
True


Other math functions

In [101]:
a = Decimal('0.1')
print(a.ln())
print(a.exp())
print(a.sqrt())

-2.302585092994045684017991455
1.105170918075647624811707826
0.3162277660168379331998893544


This will be different if we work with float values, to be more specific, let's import the math module

In [102]:
import math

In [103]:
math.sqrt(a)

0.31622776601683794

What happens is that math first converts the decimal to float

In [104]:
x = 0.01
x_dec = Decimal(0.01)

In [105]:
root_float = math.sqrt(x)
root_mixed = math.sqrt(x_dec)
root_dec = x_dec.sqrt()

In [106]:
print(f'root_float: {root_float}, root_mixed: {root_mixed}, root_dec: {root_dec}')

root_float: 0.1, root_mixed: 0.1, root_dec: 0.1000000000000000010408340856


In [107]:
print(f'root_float: {root_float ** 2}, root_mixed: {root_mixed ** 2}, root_dec: {root_dec ** 2}')

root_float: 0.010000000000000002, root_mixed: 0.010000000000000002, root_dec: 0.01000000000000000020816681712


----
## Decimals: Performance Considerations

There are some drawbacks to the Decimal class vs the float class

- not as easy to code: construction via strings or tuples
- not all mathmeatical functions that exist in the math module have a Decimal counterpart
- more memory overhead
- performance: much slower than floats (relatively)

In [108]:
from decimal import Decimal
import sys

In [115]:
a = 3.1415
b = Decimal('3.1415')

In [110]:
sys.getsizeof(a)

24

In [111]:
sys.getsizeof(b)

104

Every Decimal objects requieres larger memory

In [116]:
import time

def run_float(n=1):
    for i in range(n):
        a = 3.1415
        
def run_decimal(n=1):
    for i in range(n):
        a = Decimal('3.1415')

In [117]:
n = 10000000
start = time.perf_counter()
run_float(n)
end = time.perf_counter()
print('float: ', end-start)

float:  0.40164843100001235


In [118]:
n = 10000000
start = time.perf_counter()
run_decimal(n)
end = time.perf_counter()
print('run_decimal: ', end-start)

run_decimal:  2.4175517439998657


In [119]:


def run_float(n=1):
    a = 3.1415
    for i in range(n):
        a + a
        
def run_decimal(n=1):
    a = Decimal('3.1415')
    for i in range(n):
       a + a

In [122]:
n = 10000000
start = time.perf_counter()
run_float(n)
end = time.perf_counter()
print('float: ', end-start)
n = 10000000
start = time.perf_counter()
run_decimal(n)
end = time.perf_counter()
print('run_decimal: ', end-start)


float:  0.5072368400001324
run_decimal:  0.9023858519999521


Square Root

In [123]:
import math

def run_float(n=1):
    a = 3.1415
    for i in range(n):
        math.sqrt(a)
        
def run_decimal(n=1):
    a = Decimal('3.1415')
    for i in range(n):
       a.sqrt()
    
    
n = 5000000
start = time.perf_counter()
run_float(n)
end = time.perf_counter()
print('float: ', end-start)
n = 10000000
start = time.perf_counter()
run_decimal(n)
end = time.perf_counter()
print('run_decimal: ', end-start)


float:  0.6278716020001411
run_decimal:  40.30973671799984


Substancially slower

-----
## Complex Numbers

In [124]:
help(complex)

Help on class complex in module builtins:

class complex(object)
 |  complex(real=0, imag=0)
 |  
 |  Create a complex number from a real part and an optional imaginary part.
 |  
 |  This is equivalent to (real + imag*1j) where imag defaults to 0.
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __bool__(self, /)
 |      self != 0
 |  
 |  __divmod__(self, value, /)
 |      Return divmod(self, value).
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __float__(self, /)
 |      float(self)
 |  
 |  __floordiv__(self, value, /)
 |      Return self//value.
 |  
 |  __format__(...)
 |      complex.__format__() -> str
 |      
 |      Convert to a string according to format_spec.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getnewargs__(...)
 |  
 |  __gt__(self, value, /)
 | 

In [125]:
a = complex(1,2)
b = 1 + 2j
a == b

True

It's true in this case, but not in every scenario

In [126]:
a.real, type(a.real)

(1.0, float)

In [127]:
a.imag, type(a.imag)

(2.0, float)

In [128]:
a.conjugate()

(1-2j)

In [129]:
a = 1 + 2j
b = 10 + 8j

In [130]:
a+b

(11+10j)

In [131]:
a*b

(-6+28j)

In [132]:
a/b

(0.1585365853658537+0.07317073170731708j)

In [133]:
a**b

(0.10162518685520396-0.43309461918224995j)

In [134]:
a // 2

TypeError: can't take floor of complex number.

In [135]:
divmod(a,b)

TypeError: can't take floor or mod of complex number.

In [136]:
a = 0.1j

In [137]:
format(a.imag, '0.25f')

'0.1000000000000000055511151'

In [138]:
a + a + a == 0.3j

False

In [139]:
format((a+a+a).imag, '0.25f')

'0.3000000000000000444089210'

In [140]:
import math

In [141]:
math.sqrt(2)

1.4142135623730951

In [142]:
math.pi

3.141592653589793

In [143]:
import cmath

In [144]:
cmath.pi

3.141592653589793

In [145]:
type(cmath.pi)

float

In [146]:
a = 1 + 2j

In [147]:
math.sqrt(a)

TypeError: can't convert complex to float

In [148]:
cmath.sqrt(a)

(1.272019649514069+0.7861513777574233j)

Cmath is specifaclly made for complex numbers

In [149]:
a = 1 + 1j

In [150]:
cmath.phase(a)

0.7853981633974483

In [151]:
cmath.pi/4

0.7853981633974483

Euclidean value

In [153]:
abs(a)

1.4142135623730951

In [154]:
cmath.rect(math.sqrt(2), math.pi/4)

(1.0000000000000002+1j)

In [155]:
RHS = cmath.exp(cmath.pi * 1j)

In [156]:
cmath.isclose(RHS, 0)

False

----
## Booleans - Integer subclass

In [157]:
help(bool)

Help on class bool in module builtins:

class bool(int)
 |  bool(x) -> bool
 |  
 |  Returns True when the argument x is true, False otherwise.
 |  The builtins True and False are the only two instances of the class bool.
 |  The class bool is a subclass of the class int, and cannot be subclassed.
 |  
 |  Method resolution order:
 |      bool
 |      int
 |      object
 |  
 |  Methods defined here:
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __or__(self, value, /)
 |      Return self|value.
 |  
 |  __rand__(self, value, /)
 |      Return value&self.
 |  
 |  __repr__(self, /)
 |      Return repr(self).
 |  
 |  __ror__(self, value, /)
 |      Return value|self.
 |  
 |  __rxor__(self, value, /)
 |      Return value^self.
 |  
 |  __xor__(self, value, /)
 |      Return self^value.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create 

In [158]:
issubclass(bool,int)

True

In [159]:
type(True), id(True), int(True)

(bool, 4543016688, 1)

In [161]:
type(False), id(False), int(False)

(bool, 4543017416, 0)

In [162]:
3 < 4

True

In [163]:
type(3 < 4)

bool

In [164]:
id(3 < 4)

4543016688

This id is the same as the True id value

**True** and **False** are Singletons objects

In [165]:
None is False

False

In [166]:
1 == 2 == False
#What this statemente means, is:
1 == 2 and 2 == False

False

In [167]:
1 + True

2

In [168]:
bool(True)

True

In [169]:
bool(1)

True

---
## Booleans: Truth Values

All object in Python have an associated **truth value***

We already saw this with integers ( although to be fair, bool is a subclass of int)

But this works the same for any object

In general, the rules are straightforward

Every object has a **True** value, except:

- None
- False
- 0 in any numeric type(e.g 0, 0.0, 0+0j,...)
- empty sequences (e.g. list, tuple, string,...)
- empty mapping types(e.g. dictionary, set...)
- custom classes that implement a bool or len method that returns False or 0

In [1]:
bool(1)

True

In [2]:
bool(0)

False

Anything else that is non-zero, is **True**

In [4]:
help(bool)

Help on class bool in module builtins:

class bool(int)
 |  bool(x) -> bool
 |  
 |  Returns True when the argument x is true, False otherwise.
 |  The builtins True and False are the only two instances of the class bool.
 |  The class bool is a subclass of the class int, and cannot be subclassed.
 |  
 |  Method resolution order:
 |      bool
 |      int
 |      object
 |  
 |  Methods defined here:
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __or__(self, value, /)
 |      Return self|value.
 |  
 |  __rand__(self, value, /)
 |      Return value&self.
 |  
 |  __repr__(self, /)
 |      Return repr(self).
 |  
 |  __ror__(self, value, /)
 |      Return value|self.
 |  
 |  __rxor__(self, value, /)
 |      Return value^self.
 |  
 |  __xor__(self, value, /)
 |      Return self^value.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create 

When we do **bool(0)**, what Python is actually doing is returning this comparison: 0 != 0

In [6]:
bool(100), (100).__bool__()

(True, True)

In [7]:
bool(0), (0).__bool__()

(False, False)

In [8]:
a = []
bool(a)

False

In [9]:
a.__len__()

0

In [10]:
bool(a.__len__())

False

In [11]:
bool(0.0), bool(0+0j)

(False, False)

In [12]:
from decimal import Decimal
from fractions import Fraction

In [13]:
bool(Fraction(0,1)), bool(Decimal('0.0'))

(False, False)

In [14]:
bool(10.5), bool(1j), bool(Fraction(1,2)), bool(Decimal('2.5'))

(True, True, True, True)

Sequence types. Every sequence will be true unless the sequence is empty

In [15]:
a = []
b = ''
c = ()
bool(a), bool(b), bool(c)

(False, False, False)

In [16]:
a = [1,2]
b = 'abc'
c = (1,2)
bool(a), bool(b), bool(c)

(True, True, True)

In [18]:
a = {}
b = set()
bool(a), bool(b)

(False, False)

In [19]:
bool(None)

False

In [20]:
bool(False)

False

In [21]:
bool(True)

True

In [22]:
a = [1,2,3]
if a is not None and len(a) > 0:
    print(a[0])
else:
    print('Nothing to see here...')

1


But we can write this in a different way

In [25]:
if a:
    print(a[0])
else:
    print('Nothing to see here...')

1


---
## Booleans: Precedence and Short-Circuiting

### Precedence
not --> and --> or

In [26]:
True or True and False

True

**and** has a higher precedence than **or**. So, this is how it gets evaluated:

In [27]:
True or (True and False)

True

In [28]:
(True or True) and False

False

When we don't have the parenthesis, **and** takes the higher precedence

### Short-Circuiting
True or Y --> True
    <br>
False and Y --> False

In [29]:
a = 10
b = 2

if a/b > 2:
    print('a is at least twice b')


a is at least twice b


In [30]:
a = 10
b = 0

if a/b > 2:
    print('a is at least twice b')


ZeroDivisionError: division by zero

In [31]:
a = 10
b = 0

if b > 0:
    if a/b > 2:
        print('a is at least twice b')

Since **b>0** is false, Python did not need to evaluate the next statement

In [32]:
a = 10
b = 0

if b:
    if a/b > 2:
        print('a is at least twice b')

In [33]:
a = 10
b = 1

if b:
    if a/b > 2:
        print('a is at least twice b')

a is at least twice b


In [41]:
import string

In [42]:
help(string)

Help on module string:

NAME
    string - A collection of string constants.

DESCRIPTION
    Public module variables:
    
    whitespace -- a string containing all ASCII whitespace
    ascii_lowercase -- a string containing all ASCII lowercase letters
    ascii_uppercase -- a string containing all ASCII uppercase letters
    ascii_letters -- a string containing all ASCII letters
    digits -- a string containing all ASCII decimal digits
    hexdigits -- a string containing all ASCII hexadecimal digits
    octdigits -- a string containing all ASCII octal digits
    punctuation -- a string containing all ASCII punctuation characters
    printable -- a string containing all ASCII characters considered printable

CLASSES
    builtins.object
        Formatter
        Template
    
    class Formatter(builtins.object)
     |  Methods defined here:
     |  
     |  check_unused_args(self, used_args, args, kwargs)
     |  
     |  convert_field(self, value, conversion)
     |  
     |  format

In [44]:
a = 'c'
a in string.ascii_uppercase

False

In [48]:
string.digits

'0123456789'

In [60]:
name = 'Bob'
if name and name[0] in string.digits:
    print('Name cannot start with a digit')

---
## Booleans: Boolean Operators and Truth Values

X or Y: if X is Truth, returns X, otherwise, evaluates Y and returns it

In [61]:
'a' or [1,2]

'a'

In [62]:
'' or [1,2]

[1, 2]

In [63]:
1 or 1/0 #The right side won't get evaluated
#Otherwise, it returns an error

1

In [64]:
0 or 1/0

ZeroDivisionError: division by zero

In [65]:
s1 = None
s2 = ''
s3 = 'abc'

In [66]:
s1 = s1 or 'n/a'
s2 = s2 or 'n/a'
s3 = s3 or 'n/a'

In [67]:
s1,s2,s3

('n/a', 'n/a', 'abc')

In [68]:
[] or [0]

[0]

In [69]:
None or [0]

[0]

This is a great way to get default values

X and Y: if X is falsy, return X, otherwise, evaluate and return Y

In [71]:
print(None and 100)

None


In [72]:
print(100 and None)

None


In [73]:
[] and [0]

[]

In [75]:
a = 2 
b = 0

a/b

ZeroDivisionError: division by zero

a/b in general, but return 0 when b is zero

In [76]:
a = 2
b = 4
print(b and a/b)

0.5


In [77]:
a = 2
b = 0
print(b and a/b)

0


In [80]:
s1 = None
s2 = ''
s3 = 'abc'

print((s1 and s1[0]) or 'n/a')
print((s2 and s2[0]) or 'n/a')
print(s3 and s3[0])

n/a
n/a
a


In [81]:
help(bool)

Help on class bool in module builtins:

class bool(int)
 |  bool(x) -> bool
 |  
 |  Returns True when the argument x is true, False otherwise.
 |  The builtins True and False are the only two instances of the class bool.
 |  The class bool is a subclass of the class int, and cannot be subclassed.
 |  
 |  Method resolution order:
 |      bool
 |      int
 |      object
 |  
 |  Methods defined here:
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __or__(self, value, /)
 |      Return self|value.
 |  
 |  __rand__(self, value, /)
 |      Return value&self.
 |  
 |  __repr__(self, /)
 |      Return repr(self).
 |  
 |  __ror__(self, value, /)
 |      Return value|self.
 |  
 |  __rxor__(self, value, /)
 |      Return value^self.
 |  
 |  __xor__(self, value, /)
 |      Return self^value.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create 

In [82]:
bool('abc')

True

In [83]:
bool('')

False

In [84]:
not bool('abc')

False

In [85]:
not bool('')

True

Is not necessary to use **bool** instance

In [86]:
not 'abc'

False

In [87]:
not None

True

In [88]:
not 0

True

----
## Comparison Operators

### Categories of Comparison Operators

- Binary operator. It takes two operands to accomplish the operation.

- Evaluate to a **bool** value

+ Identitity Operations -- > **is** **is not**

+ Value Comparisons -- >   **==** **!=**

+ Ordering Comparisons -- > **<** **<=** **>** **>=**

+ Membership Operations -- > **in** **not in**


### Numeric Types

Value comparison will work with all numeric types

Mixed types (except complex) in value and ordering comparisons is supported

Note: Value equality operators work between floats and Decimals, but as we have seen before, using value equality with floats has some issues!

10.0 == Decimal('10.0') -> ***True***

0.1 == Decimal('0.1') -> ***False***
.
.
.


### Ordering Comparisons

Again, these work across all numeric types, except for complex numbers

1 < 3.14 -> ***True***

Fraction(22,7) -> math.pi -> ***True***

.

.

.

### Chained Comparisons

a == b == c **->** a == b and b == c

a < b < c **->** a < b and b < c

1 == Decimal('1.0') == Fraction(1,1) -> ***True***

1 == Decimal('1.5') == Fraction(3,2) -> ***False***

5 < 6 > 2 ->  5 < 6 and 6 > 2 -> ***True***

In [97]:
0.1 is (3+4j)

<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<ipython-input-97-aadb6322e9fe>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  0.1 is (3+4j)


False

In [96]:
3 is 3

<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<ipython-input-96-1ee75a7b8cc8>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  3 is 3


True

In [98]:
[1,2] is [1,2]

False

In [99]:
'a' in 'this is a test'

True

In [100]:
3 in [1,2,3]

True

In [101]:
3 not in [1,2,3]

False

In [102]:
3 < 5

True

In [104]:
from decimal import Decimal
from fractions import Fraction

4 < Decimal('10.5')

True

In [107]:
4 == 4 + 0j

True

In [108]:
1 < 2 < 3

True

In [109]:
1 < 2 and 2 < 3

True

In [110]:
import string

In [111]:
'A' < 'a'

True

In [116]:
'A' < 'a' < 'z' > 'Z' in  string.ascii_letters

True

In [114]:
'2' < '1'

False